src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs

```diff
@@ -40,32 +41,46 @@ public PipeCompletionCallbacks TryComplete(Exception exception = null)
         }
 
         public PipeCompletionCallbacks AddCallback(Action<Exception, object> callback, object state)
+        {
+            if (_callbackCount == 0)
+            {
+                _firstCallback = new PipeCompletionCallback(callback, state);
+                _callbackCount++;
+            }
+            else
+            {
+                EnsureSpace();
+
+                // -1 to adjust for _firstCallback
+                var callbackIndex = _callbackCount - 1;
+                _callbackCount++;
+                _callbacks[callbackIndex] = new PipeCompletionCallback(callback, state);
+            }
+
+            if (IsCompleted)
+            {
+                return GetCallbacks();
+            }
+
+            return null;
+        }
+
+        private void EnsureSpace()
         {
             if (_callbacks == null)
             {
                 _callbacks = s_completionCallbackPool.Rent(InitialCallbacksSize);
             }
 
-            int newIndex = _callbackCount;
-            _callbackCount++;
+            int newLength = _callbackCount - 1;
 
-            if (newIndex == _callbacks.Length)
+            if (newLength == _callbacks.Length)
```

Comment:
`s_completionCallbackPool.Rent(_callbacks.Length * 2);` has potential for integer over flow. How likely is it to hit the overflow (i.e. callbacks.Length > 1 GB)? Check and throw?

This way you also avoid the risk of overflowing when doing `_callbackCount++;`  above.

Reply:
It's extremely unlikely.